### PR TITLE
Add a Gemini models for text and multi-modal.

### PR DESCRIPTION
**Pull request by markmcd**
# Description

Includes each of {text, mm} x {completion, chat} x {stream, one-shot} x {sync, async} for Gemini's Developer API.

And two notebooks - regular text & multi-modal.

*NOTE*: This will not work until Gemini-supported SDKs are published.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by hatianzhang** [comment]
One thing I need to update is including Gemini for https://docs.llamaindex.ai/en/stable/module_guides/models/llms.html.

Will do after Gemini Release so that we can do the testing with llamaindex

File: docs/examples/multi_modal/gemini.ipynb [hunk 1]
```diff
@@ -0,0 +1,319 @@
+{
```
Comment (hatianzhang) [inline docs/examples/multi_modal/gemini.ipynb:1]: <p>do we need to specify a version for <code>google-generativeai</code> pkg </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9452/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (markmcd) [reply docs/examples/multi_modal/gemini.ipynb:1]: we do (0.3.0) - but the package is published now, and the constraints are the same. can't hurt to add it though.

Comment (markmcd) [reply docs/examples/multi_modal/gemini.ipynb:1]: done

**Comment by hatianzhang** [comment]
@markmcd one thing is confirm is that Gemini `gemini-pro-vision` does not support `input: text/image` and `output: image` atm right?

**Comment by hatianzhang** [comment]
@markmcd feel free to check [docs/examples/multi_modal/gemini.ipynb](https://github.com/run-llama/llama_index/pull/9452/files#diff-5f74c69b91425556eeb03ac013595e1e7f2cbbf829dfaf647bc5a5c8a765cde4), which is a simple RAG I just added for combining Gemini + Llamaindex

LMK if you have any questions or concerns/

**Comment by markmcd** [comment]
> @markmcd one thing is confirm is that Gemini `gemini-pro-vision` does not support `input: text/image` and `output: image` atm right?

Right - this first stage is text-output only.

> feel free to check docs/examples/multi_modal/gemini.ipynb

Amazing! That's impressive.

**Comment by hatianzhang** [comment]
@markmcd I also have a draft PR for Gemini Embedding.
I might push it tmrw. But one thing need to know is that "what is the difference between `title` and `content`

example code from Gemini API doc
```
result = genai.embed_content(model="models/embedding-001",
                                content="This is some text from the doc",
                                task_type="retrieval_document",
                                title="this is a document title")
```

**Comment by hatianzhang** [comment]
btw @markmcd the Gemini Embedding PR https://github.com/run-llama/llama_index/pull/9476. need some polish and tests

File: docs/examples/multi_modal/gemini.ipynb [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
```
Comment (jerryjliu) [inline docs/examples/multi_modal/gemini.ipynb:1]: <p>nit: </p><ul><li>"First, We also show several functions" --&gt; "First, we show several functions that we now support" </li><li>Google Map -&gt; Google Maps </li><li>"Using the gemini-pro model synthesizes the results" --&gt; "Use the gemini-pro model to synthesize the results and recommend restaurants..." </li><li><br></li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9452/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (jerryjliu) [inline docs/examples/multi_modal/gemini.ipynb:1]: <ul><li>"Structured Output Parser from Image" --&gt; "Structured Output Parsing from an Image" or "Structured Data Extraction from an Image"</li></ul><p> </p><p>I would also add a description here </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9452/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (jerryjliu) [inline docs/examples/multi_modal/gemini.ipynb:1]: <ul><li>"perfectly generate" --&gt; "perfectly generates "</li><li>"also recognize" --&gt; "also recognizes" </li><li>Google Map --&gt; Google Maps </li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9452/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (jerryjliu) [inline docs/examples/multi_modal/gemini.ipynb:1]: <p>Better Title: "Build Multi-Modal RAG for Restaurant Recommendation". Sub-text can be "Our stack consists of Gemini + LlamaIndex + Pydantic structured output capabilities" </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9452/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (jerryjliu) [inline docs/examples/multi_modal/gemini.ipynb:1]: <p>*Use Gemini to synthesize the results and recommend...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9452/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

**Review by jerryjliu — APPROVED** [review]
@hatianzhang i think we should clean up the grammar/spelling nits above, but approving  to unblock

**Pull request merged**